charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+		f := h.relationType.Field(i)
+		registeredRelationsMutex.Lock()
+		reg := registeredRelations[f.Type]
+		registeredRelationsMutex.Unlock()
+		if reg == nil {
+			// This shouldn't be able to happen because we should have
+			// checked that all fields have registered types during Register.
+			panic(errgo.Newf("no registered relation for type %s in field %s", f.Type, f.Name))
+		}
+		args[0] = srv.relationState.Field(0).Addr()
+
+		// Create an instance of type D and unmarshal the data into it.
+		args[1] = reflect.New(reg.d)
+		data := state.RelationValues[f.Name]
+		if err := json.Unmarshal(data, args[1].Interface()); err != nil {
+			return errgo.Notef(err, "cannot unmarshal relation data into field %s (type %v, data %q)", f.Name, args[1].Type(), data)
+		}
+		args[1] = args[1].Elem()
+		// TODO we could potentially call all these in parallel
+		// so that any network connections they might be
+		// making would be concurrent.
+		err, _ := reg.setFieldValue.Call(args)[0].Interface().(error)
+		if err != nil {
+			if errgo.Cause(err) == ErrRestartNeeded {
+				restartNeeded = true
+			} else {
+				return errgo.Notef(err, "cannot set relation field %s (type %s)", f.Name, f.Type)
+			}
+		}
+	}
+	if restartNeeded {
+		return ErrRestartNeeded
+	}
+	return nil
+}
+
+// closeResources closes any current listeners. Called
+// with srv.mu held.
+func (srv *server) closeResources() {
+	if srv.handler != nil {
+		srv.handler.Close()
+		srv.handler = nil
+	}
+	if srv.httpListener != nil {
+		srv.httpListener.Kill()
+		srv.httpListener.Wait()
+		srv.httpListener = nil
+	}
+	if srv.httpsListener != nil {
+		srv.httpsListener.Kill()
+		srv.httpsListener.Wait()
+		srv.httpsListener = nil
+	}
+}
+
+func (srv *server) saveState(state ServerState) error {
+	data, err := json.Marshal(state)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	path := srv.statePath()
+	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errgo.Mask(err)
+	}
+	if err := ioutil.WriteFile(path, data, 0600); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (srv *server) loadState() (ServerState, error) {
+	data, err := ioutil.ReadFile(srv.statePath())
+	if os.IsNotExist(err) {
+		return ServerState{}, nil
+	}
+	if err != nil {
+		return ServerState{}, errgo.Mask(err)
+	}
+	var state ServerState
+	if err := json.Unmarshal(data, &state); err != nil {
+		return ServerState{}, errgo.Notef(err, "cannot unmarshal state %q", data)
+	}
+	return state, nil
+}
+
+func (srv *server) statePath() string {
+	return filepath.Join(srv.stateDir, "serverstate.json")
+}
+
+// tcpKeepAliveListener is stolen from net/http
+
+// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
+// connections. It's used by ListenAndServe and ListenAndServeTLS so
+// dead TCP connections (e.g. closing laptop mid-download) eventually
+// go away.
+type tcpKeepAliveListener struct {
+	*net.TCPListener
+}
+
+func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
```

Comment:
Might be worth a godoc stolen from net/http but also stating again that it's stolen from there